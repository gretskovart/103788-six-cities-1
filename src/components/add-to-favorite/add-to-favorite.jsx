import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {withRouter} from 'react-router-dom';
import favoritePost from './../../api/favorite';

import {actionCreators} from '../../reducer/data/data';

class AddToFavorite extends PureComponent {
  constructor(props) {
    super(props);

    this._addToFavorite = this._addToFavorite.bind(this);
  }

  render() {
    const {isFavorite, isUserAuthorized, id} = this.props;
    const isFavoriteClassName = isFavorite ? `place-card__bookmark-button--active` : ``;

    return (
      <button
        className={`place-card__bookmark-button button ${isFavoriteClassName}`}
        type="button"
        onClick={() => this._addToFavorite(isUserAuthorized, id, isFavorite)}
      >
        <svg className="place-card__bookmark-icon" width="18" height="19">
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }

  _addToFavorite(isUserAuthorized, id, isFavorite) {
    const {history, addToFavorite} = this.props;
    const favoritAction = isFavorite ? 0 : 1;

    if (!isUserAuthorized) {
      history.push(`/login`);
    } else {
      favoritePost(id, favoritAction);
      addToFavorite(id);
    }
  }
}

const mapStateToProps = (state) => {
  return {
    isUserAuthorized: state.user.isUserAuthorized
  };
};

const mapDispatchToProps = (dispatch) => ({
  addToFavorite: (id) => {
    dispatch(actionCreators.addToFavorite(id));
  }
});

export {AddToFavorite};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(AddToFavorite));


AddToFavorite.propTypes = {
  isFavorite: PropTypes.bool,
  isUserAuthorized: PropTypes.bool.isRequired,
  history: PropTypes.object,
  id: PropTypes.number.isRequired,
  addToFavorite: PropTypes.func
};
