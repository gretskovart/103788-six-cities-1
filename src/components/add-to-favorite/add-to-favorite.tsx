import * as React from 'react';
import {PureComponent} from 'react';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import favoritePost from '../../api/favorite';
import {constants} from '../../helpers';
import {actionCreators} from '../../reducer/data/data';

interface Props {
  isFavorite: boolean;
  isUserAuthorized: boolean;
  id: number;
  property: string;
  history: {};
  addToFavorite: (id: number) => void
}

class AddToFavorite extends PureComponent<Props> {
  constructor(props: Props) {
    super(props);

    this._addToFavorite = this._addToFavorite.bind(this);
  }

  render() {
    const {isFavorite, isUserAuthorized, id, property} = this.props;
    const propertyClassName: string = property ? property : `place-card`;
    const isFavoriteClassName: string = isFavorite ? `${propertyClassName}__bookmark-button--active` : ``;
    const {FAVORITE_BTN} = constants;
    const typeBtn: BtnSize = property ? FAVORITE_BTN.DETAIL_PAGE : FAVORITE_BTN.LIST_PAGE;

    interface BtnSize {
      width: number,
      height: number
    }

    return (
      <button
        className={`${propertyClassName}__bookmark-button button ${isFavoriteClassName}`}
        type="button"
        onClick={() => this._addToFavorite(isUserAuthorized, id, isFavorite)}
      >
        <svg className="place-card__bookmark-icon" width={typeBtn.width} height={typeBtn.height}>
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }

  _addToFavorite(isUserAuthorized: boolean, id: number, isFavorite: boolean): void {
    const {history, addToFavorite} = this.props;
    const favoritAction = isFavorite ? 0 : 1;

    if (!isUserAuthorized) {
      history.push(`/login`);
    } else {
      favoritePost(id, favoritAction);
      addToFavorite(id);
    }
  }
}

const mapStateToProps = (state) => {
  return {
    isUserAuthorized: state.user.isUserAuthorized
  };
};

const mapDispatchToProps = (dispatch): {addToFavorite: Function} => ({
  addToFavorite: (id: number) => {
    dispatch(actionCreators.addToFavorite(id));
  }
});

export {AddToFavorite};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(withRouter(AddToFavorite));
