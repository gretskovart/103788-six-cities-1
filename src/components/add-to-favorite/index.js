import AddToFavorite from './add-to-favorite.tsx';

export default AddToFavorite;
