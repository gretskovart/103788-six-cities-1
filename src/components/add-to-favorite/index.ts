import AddToFavorite from './add-to-favorite';

export default AddToFavorite;
