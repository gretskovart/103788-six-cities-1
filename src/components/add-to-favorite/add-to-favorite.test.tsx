import * as React from 'react';
import {AddToFavorite} from './add-to-favorite';
import renderer from 'react-test-renderer';

const mock = {
  isFavorite: true,
  isUserAuthorized: true,
  id: 1,
  property: `property`,
  history: {}
};

it(`<AddToFavorite /> renders correctly`, () => {
  const {isFavorite, isUserAuthorized, id, property, history} = mock;

  const tree = renderer
    .create(
        <AddToFavorite
          isFavorite={isFavorite}
          isUserAuthorized={isUserAuthorized}
          id={id}
          property={property}
          history={history}
          addToFavorite={jest.fn()}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
