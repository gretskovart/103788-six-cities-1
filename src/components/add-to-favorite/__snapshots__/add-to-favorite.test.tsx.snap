// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToFavorite /> renders correctly 1`] = `
<button
  className="property__bookmark-button button property__bookmark-button--active"
  onClick={[Function]}
  type="button"
>
  <svg
    className="place-card__bookmark-icon"
    height={33}
    width={31}
  >
    <use
      xlinkHref="#icon-bookmark"
    />
  </svg>
  <span
    className="visually-hidden"
  >
    To bookmarks
  </span>
</button>
`;
